### PR TITLE
Add capability to redirect to Python logger

### LINKED ISSUE #33215: Redirect log messages to Python logger
When using Mantid from Python, it would be useful if log messages were sent to Python's loggers instead of printing to terminal or file.

This would allow users to control if and how logs are shown and also combine them with logs from other software.
It should be fairly straight forward to configure Mantid's logger during import from Python to forward all messages to the Python logger.

I have already started looking into it because it would help us a lot in scipp. But I still wanted to inform you before submitting a PR out of the blue.
Do you have any comments and do you even think this is worth adding to Mantid?

### PR DESCRIPTION
**Pull request by jl-wynen**
**Description of work.**

Defines a new `Poco::Channel` which sends messages to a logger from Python's `logging` framework and `log_to_python` to help set up the logger.

The Python logger has fewer levels than Poco's. It is possible to define new levels in Python but those might clash with user defined levels. Given that the new mechanism is meant to be used in combination with other packages that might log, defining new levels could cause problems. This PR, therefore,  maps several Poco levels onto the same Python level.

It is unclear to me how this should interact with the rest of Mantid's logging config. This PR does not break any existing behaviour and leaves it up to the user to configure everything as they need it.

**To test:**
```.py
import logging
from mantid.utils.logging import log_to_python
from mantid.kernel import logger

logging.basicConfig(format='PYTHON-LOGGER:: <%(name)s> %(levelname)s : %(message)s',
                    level=logging.INFO)
log = logging.getLogger()

log_to_python()

logger.information('[[info]]')
logger.warning('[[warning]]')
logger.error('[[error]]')
logger.fatal('[[fatal]]')
```
should print
```
PYTHON-LOGGER:: <Mantid> INFO : [[info]]
PYTHON-LOGGER:: <Mantid> WARNING : [[warning]]
PYTHON-LOGGER:: <Mantid> ERROR : [[error]]
PYTHON-LOGGER:: <Mantid> CRITICAL : [[fatal]]
```

See also new `loggingTest.test_log_to_python`.

Fixes #33215

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by jl-wynen** [comment]
Comments in the issue by @peterfpeterson 
> and am not sure that log_to_python() is a necessary function rather than instructing users on how to modify their properties file. 

I added it as a convenience, especially because it depends on the order in which you set the keys. But it is not a requirement to make the logger usable.

> On linux, there is a system level properties file that could set the logging channel for users on that machine as a whole.

Using a system or even user wide setting would mean that it is used by both the GUI and Python package. This seems like a bad idea to me.

**Comment by peterfpeterson** [comment]
Bootstrapping the logging system is a pain. Current workbench and friends can't start logging until configservice and logging are stood up anyhow. Since python is integral to mantid working, I don't think it is a huge issue if you have to have python up to log.

File: Framework/PythonInterface/core/src/PythonLoggingChannel.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  Mantid::PythonInterface::GlobalInterpreterLock gil;
+  auto logger = (boost::python::import("logging").attr("getLogger")("Mantid"));
+  m_pyLogger = std::make_unique<boost::python::object>(std::move(logger));
+}
+
+// The special behavior here is needed because Poco's LoggingFactory can be destroyed
+// after the Python interpreter was shut down.
+PythonLoggingChannel::~PythonLoggingChannel() {
+  if (Py_IsInitialized()) {
+    Mantid::PythonInterface::GlobalInterpreterLock gil;
```
Comment (peterfpeterson) [inline Framework/PythonInterface/core/src/PythonLoggingChannel.cpp:61]: I may be reading this wrong, but I think `release` should happen if python is initialized and setting it to `nullptr` should happen if python isn't initialized.

Comment (jl-wynen) [reply Framework/PythonInterface/core/src/PythonLoggingChannel.cpp:61]: `m_pyLogger = nullptr;` makes `std::unique_ptr` delete the contained object which calls its destructor which in turn needs to access Python.
The `release` hack allows us to only free the memory allocated by us without calling into Python via `*m_pyLogger`'s destructor.

File: Framework/PythonInterface/test/python/mantid/utils/loggingTest.py [hunk 2]
```diff
@@ -7,14 +7,34 @@
...
+    def __init__(self):
+        super().__init__(level=logging.DEBUG)
+        self.records = []
+
+    def emit(self, record: logging.LogRecord):
+        self.records.append(record)
+
+
+@contextmanager
+def backup_config():
```
Comment (peterfpeterson) [inline Framework/PythonInterface/test/python/mantid/utils/loggingTest.py:28]: It feels like there is a better place to put this. I've written this same functionality myself a bunch of times which isn't very DRY.

File: dev-docs/source/Logging.rst [hunk 3]
```diff
@@ -126,6 +126,26 @@ In general:
   from mantid.kernel import logger
   logger.warning('this is a custom warning')
 
+Using :code:`logging` in Python
```
Comment (peterfpeterson) [inline dev-docs/source/Logging.rst:129]: Can you add a link to/form `docs/source/api/python/mantid/kernel/Logger.rst` in this as well. That documentation is auto-generated and not nearly as useful.

File: docs/source/release/v6.4.0/framework.rst [hunk 4]
```diff
@@ -21,6 +21,8 @@ Data Objects
 Python
 ------
 
+- Added possibility to forward log messages to Python, see `mantid.utils.logging.log_to_python`.
```
Comment (peterfpeterson) [inline docs/source/release/v6.4.0/framework.rst:24]: Please make this a link to `dev-docs/source/Logging.rst`

**Review by SimonHeybrock — DISMISSED** [review]
Nice, looks good to me, aside from Pete's comments. 👍

File: docs/source/release/v6.4.0/framework.rst (hunk 4 above)
Comment (jl-wynen) [reply docs/source/release/v6.4.0/framework.rst:24]: How do you link from the user to the dev docs? Is there a nicer way than using URLs to the published online documentation?

Comment (peterfpeterson) [reply docs/source/release/v6.4.0/framework.rst:24]: It should be
```
:ref:`mantid.utils.logging.log_to_python`
```

Comment (jl-wynen) [reply docs/source/release/v6.4.0/framework.rst:24]: Ah, ok.
But I was also asking about your other comment where I need to link to  a file / section.

File: Framework/PythonInterface/test/python/mantid/utils/loggingTest.py (hunk 2 above)
Comment (jl-wynen) [reply Framework/PythonInterface/test/python/mantid/utils/loggingTest.py:28]: How about
```.py
config = ConfigService.Instance()
with config.temporary():
    config[key] = value
    # do stuff
# config gets reset here
```
But I don't know of the larger implications here. Would backing up and restoring the entire config re-init important components and cause problems?

Or are you referring to the logging config specifically?

Comment (peterfpeterson) [reply Framework/PythonInterface/test/python/mantid/utils/loggingTest.py:28]: I meant as a utility for testing, not specific to logging. If you can think of a good place to put this so we can advertise it to developers I'd be very appreciative. However, it is outside the scope of this PR so I won't block merging based on this.

File: dev-docs/source/Logging.rst (hunk 3 above)
Comment (peterfpeterson) [reply dev-docs/source/Logging.rst:129]: The link here should be
```rst
:class:`mantid.kernel.Logger`
```

Comment (jl-wynen) [reply dev-docs/source/Logging.rst:129]: Ok. I thought you meant to link the other way round.

File: Framework/PythonInterface/test/python/mantid/utils/loggingTest.py (hunk 2 above)
Comment (jl-wynen) [reply Framework/PythonInterface/test/python/mantid/utils/loggingTest.py:28]: Looks like `Framework/PythonInterface/test/testhelpers` is a good place. Could be either in `__init__.py` or in a new separate module.

**Review by peterfpeterson — DISMISSED** [review]

**Review by peterfpeterson — DISMISSED** [review]

**Comment by jl-wynen** [comment]
I don't know what is wrong with the failing test. It looks like a problem with QT and I don't see how this PR caused it. Can someone help?

**Comment by SimonHeybrock** [comment]
A similar issue was reported in #unreliable-tests a while back, I am trying to restart your build for now.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**